src/Common/MemoryBufferWriter.cs

```diff
@@ -170,5 +211,65 @@ public byte[] ToArray()
 
             return result;
         }
+
+        public override void Flush() { }
+        public override Task FlushAsync(CancellationToken cancellationToken) => Task.CompletedTask;
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        public override void WriteByte(byte value)
+        {
+            if (_currentSegment != null && _position < _currentSegment.Length)
+            {
+                _currentSegment[_position] = value;
+            }
+            else
+            {
+                EnsureCapacity(1);
+                _currentSegment[_position] = value;
+            }
+
+            _bytesWritten++;
+            _position++;
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            if (_currentSegment != null && _position + count < _currentSegment.Length)
```

Comment:
Use `_position` lots so take to local/register so it isn't always reading from stack memory
```csharp
var position = _position;
if (_currentSegment != null && position < _currentSegment.Length - count)
{
    Buffer.BlockCopy(buffer, offset, _currentSegment, position, count);
    _position = position + count;
```